### ISSUE TITLE
[MASSEMBLY-644] Exclusions in dependencySet inside moduleSet->binaries have no effect

### ISSUE THREAD

**Issue by jira-importer** [apache/maven-assembly-plugin#804]
**[Maik Riechert](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=maik.riechert@arcor.de)** opened **[MASSEMBLY-644](https://issues.apache.org/jira/browse/MASSEMBLY-644?redirect=false)** and commented

I'm trying to create a distribution assembly of a multi-module project in a separate distribution module which works quite nice, except that exclusions of dependencies of modules seem to be ignored. This is strange, as the same exclusions do work in an assembly inside a submodule.

This is the problematic descriptor:

```
<?xml version="1.0" encoding="UTF-8"?>
<assembly>

	<id>lwjgl</id>
	<formats>
		<format>zip</format>
	</formats>
	<includeBaseDirectory>false</includeBaseDirectory>

	<moduleSets>
		<moduleSet>
			<useAllReactorProjects>true</useAllReactorProjects>

			<includes>
				<include>com.ardor3d:ardor3d-animation</include>
				<include>com.ardor3d:ardor3d-awt</include>
				<include>com.ardor3d:ardor3d-collada</include>
				<include>com.ardor3d:ardor3d-core</include>
				<include>com.ardor3d:ardor3d-effects</include>
				<include>com.ardor3d:ardor3d-extras</include>
				<include>com.ardor3d:ardor3d-lwjgl</include>
				<include>com.ardor3d:ardor3d-math</include>
				<include>com.ardor3d:ardor3d-savable</include>
				<include>com.ardor3d:ardor3d-swt</include>
				<include>com.ardor3d:ardor3d-terrain</include>
				<include>com.ardor3d:ardor3d-ui</include>
			</includes>

			<binaries>
				<unpack>false</unpack>
				<dependencySets>
					<dependencySet>
						<excludes>
							<exclude>*:lwjgl*natives-*</exclude>
							<exclude>*:jinput*natives-*</exclude>
						</excludes>
					</dependencySet>
				</dependencySets>
			</binaries>
		</moduleSet>
	</moduleSets>

	<fileSets>
		<fileSet>
			<directory>target/natives</directory>
			<outputDirectory>natives</outputDirectory>
			<excludes>
				<exclude>*jogl*</exclude>
				<exclude>*nativewindow*</exclude>
				<exclude>*newt*</exclude>
				<exclude>*gluegen*</exclude>
				<exclude>META-INF/</exclude>
			</excludes>
		</fileSet>
	</fileSets>
</assembly>
```

This assembly descriptor can be found here: https://github.com/neothemachine/Ardor3D/blob/assembly/ardor3d-distribution/assembly-lwjgl.xml

It produces a zip which also contains "lwjgl-platform-2.8.4-natives-linux.jar" and others which should be matched by the filter.

The submodule where the exclusions work can be found at: https://github.com/neothemachine/Ardor3D/tree/assembly/ardor3d-examples

If you need any other information, please tell me.

---

**Affects:** 2.4

**Comment by jira-importer** [comment]
**[Maik Riechert](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=maik.riechert@arcor.de)** commented

Sorry, made a mistake on formatting and can't edit it. The bold parts are surrounded by stars...

Edit: The debug output I get is the following:

```
[DEBUG] Processing binary dependencies for module project: com.ardor3d:ardor3d-lwjgl:bundle:0.9-SNAPSHOT
[DEBUG] Processing DependencySet (output=null)
[DEBUG] Filtering dependency artifacts WITHOUT transitive dependency path information.
[DEBUG] Statistics for Excludes filter:
o '*:lwjgl*natives-*'
o '*:jinput*natives-*'

[WARNING] The following patterns were never triggered in this artifact exclusion filter:
o  '*:lwjgl*natives-*'
o  '*:jinput*natives-*'
[...]
[DEBUG] Adding dependency artifact org.lwjgl.lwjgl:lwjgl-platform:jar:natives-linux:2.8.4.
[...]
```

Shouldn't this artifact be matched by ```
**:lwjgl**natives-*

```?

As mentioned, it works for an assembly inside a module (ardor3d-examples):

```

[DEBUG] Filtering dependency artifacts WITHOUT transitive dependency path information.
[DEBUG] org.lwjgl.lwjgl:lwjgl-platform:jar:natives-windows:2.8.4 was removed by one or more filters.
[DEBUG] org.lwjgl.lwjgl:lwjgl-platform:jar:natives-linux:2.8.4 was removed by one or more filters.
[DEBUG] org.lwjgl.lwjgl:lwjgl-platform:jar:natives-osx:2.8.4 was removed by one or more filters.
[DEBUG] net.java.jinput:jinput-platform:jar:natives-linux:2.0.5 was removed by one or more filters.
[DEBUG] net.java.jinput:jinput-platform:jar:natives-windows:2.0.5 was removed by one or more filters.
[DEBUG] net.java.jinput:jinput-platform:jar:natives-osx:2.0.5 was removed by one or more filters.
[DEBUG] Statistics for Excludes filter:
o '**:lwjgl**natives-\*'
o '**:jinput**natives-*'

[DEBUG] The following artifacts were removed by this artifact exclusion filter:
org.lwjgl.lwjgl:lwjgl-platform:jar:natives-windows:2.8.4
org.lwjgl.lwjgl:lwjgl-platform:jar:natives-linux:2.8.4
org.lwjgl.lwjgl:lwjgl-platform:jar:natives-osx:2.8.4
net.java.jinput:jinput-platform:jar:natives-linux:2.0.5
net.java.jinput:jinput-platform:jar:natives-windows:2.0.5
net.java.jinput:jinput-platform:jar:natives-osx:2.0.5

```
```

**Comment by jira-importer** [comment]
**[Maik Riechert](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=maik.riechert@arcor.de)** commented

Alright... I found the problem. The module was using the (super pom) default plugin version (2.2-beta5) while the problematic assembly uses the current 2.4. Apparently, the behaviour of pattern matching has changed, which is implemented in maven-common-artifact-filters (2.2-beta5 used version 1.1, while 2.4 uses 1.4) in the method org.apache.maven.shared.artifact.filter.PatternIncludesArtifactFilter#matchAgainst. The docs say ```
Additionally, wildcards can be used, as in **:maven-**

```which in my opinion is not exact enough, given the specific behaviour of the implementation.

Changing my patterns to
```

\<exclude>**:lwjgl**:**:natives-**\</exclude>
\<exclude>**:jinput**:**:natives-**\</exclude>

```
makes it work in 2.4.

If I understand the code correctly, then patterns must be given in a form "x:y:..." where each additional ":..." segment is optional as long as nothing else follows it. There's also some special case code which tries to handle the other direction so that patterns can have the form "*:y:z" which means that only the last two segments y and z are matched and must in fact be the end, and anything before that doesn't matter. For this special case it is important which id-forms get checked against, namely the shortId "groupId:artifactId", an id without version "groupId:artifactId:type[:classifier]", and the fully qualified id "groupId:artifactId:type[:classifier]:version". The fact that the classifier is left out in the string segments to match against is a bit troubling to me when reasoning about the matching logic, but anyway.
```

**Comment by jira-importer** [comment]
**[Karl Heinz Marbaise](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=khmarbaise)** commented

The [documenation describes](http://maven.apache.org/plugins/maven-assembly-plugin/assembly.html#class_moduleSet) the format as follows:

> `groupId:artifactId` or may be fully qualified in the
> form `groupId:artifactId:type\[:classifier\]:version`.
> Additionally, wildcards can be used, as in **:maven-**

BTW: Where did you find those things in the code? Can you give some detailed locations?

Would you suggest to improve the docs? Yes than it would be nice to open a separate JIRA issue. Apart from that i assume based on your explanations this is not a bug so i can close the issue. If you have further information or any objections please don't hesitate to reopen the issue.
